concordium-std/src/impls.rs

```diff
@@ -680,19 +727,23 @@ where
     fn next(&mut self) -> Option<Self::Item> {
         self.state_iter.next().and_then(|mut entry| {
```

Comment:
This is where we should get the key from the `state_iter` (mentioned previously).

I am not sure what the agreement is regarding `state_iterator_key_read`.
From `prims.rs` it seems like it should be called _before_ `next()` to get the correct key.
But if that is the case, then it `state_iterator_key_read` should also be able to return an error, if there is no next. (Or we should peek first). What do you think @abizjak ?

Reply:
Yes, you are right. `state_iterator_key_read` should return an Option indicating whether there is a next value or not.
I'll revise the host implementation and specification.

This also makes it cheaper to use to check whether an iterator is empty (i.e., `is_empty` method above for the state map).